core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -70,18 +70,206 @@ class SYCLInternal {
 
   std::unique_ptr<sycl::queue> m_queue;
 
+  // USMObjectMem is a reusable buffer for a single object
+  // in USM memory
+  template <sycl::usm::alloc Kind>
+  class USMObjectMem {
+   public:
+    class Deleter {
+     public:
+      Deleter() = default;
+      explicit Deleter(USMObjectMem* mem) : m_mem(mem) {}
+
+      template <typename T>
+      void operator()(T* p) const noexcept {
+        assert(m_mem);
+        assert(sizeof(T) == m_mem->m_size);
+
+        if constexpr (sycl::usm::alloc::device == kind)
+          // Only skipping the detor on trivially copyable types
+          static_assert(std::is_trivially_copyable_v<T>);
+        else
+          p->~T();
+
+        m_mem->m_size = 0;
+      }
+
+     private:
+      USMObjectMem* m_mem = nullptr;
+    };
+
+    static constexpr sycl::usm::alloc kind = Kind;
+
+    void reset() {
+      assert(!m_size);
+
+      if (m_data) {
+        sycl::free(m_data, *m_q);
+        m_capacity = 0;
+        m_data     = nullptr;
+      }
+      m_q.reset();
+    }
+
+    void reset(sycl::queue q) {
```

Comment:
What does this do? Reassociating the object with a different queue? After deleting all its stuff?

Reply:
When the `SYCLInternal` object is first constructed, the `USMObjectMem` objects it contains (currently `m_indirectKernelMem` and `m_reductionResultMem`) are default constructed and do not have a sycl::queue associated with them, as we don't as yet have one to give to them.

When we get the queue, we need some way to set it in `m_indirectKernelMem`, etc.  Previous versions just assigned over the old `USMObjectMem` object (`m_indirectKernelMem = IndirectKernelMem(q);`), but assignment of one `sycl::queue` to another (which it did internally) is not working under SYCL+Cuda.  This leaves either changing `m_indirectKernelMem` from being of type `IndirectKernelMem` to type `optional<IndirectKernelMem>` (which requires all uses to dereference it first) or to add a `reset(q)` method.  It is modeled on the `shared_ptr`/`unique_ptr` `.reset()` method.

`reset(...)` is just maintaining the relationship that `m_data` points to memory allocated with `m_q`, so if the old queue goes away, we have to free whatever memory was already allocated first.  While I expect that `reset(q)` is called exactly once (and `m_data` can only be set after `m_q` is engaged, so the first `reset(q)` call doesn't call `sycl::free`), I didn't see any reason to preclude the possibility that it gets called more than once in the future.